### DISCUSSION TITLE
Reduce boilerplates and improve ergonomics

### DISCUSSION THREAD

**Discussion by Multirious** [Multirious/bevy_tween#15] [General]
Currently the core APIs is designed around ECS so we can't do anything about it. Using `commands.spawn()` purely would be an ergonomics hell for complex animations. We can mitigate this by using builders, shortcuts, and helpers which is implemented in the crate. I'd like to discuss further and gather up ideas to improve this through this discussion, join in if you have any ideas or extra information!

**Comment by Multirious** [comment]
Strip off the `Span` prefix from span tweens https://github.com/Multirious/bevy_tween/issues/11#issue-2242168374

**Comment by Multirious** [comment]
# What can be done by you, right now

This is a general guideline to improve the ergonomics utilizing language features already provided by Rust and exploit the flexibility of this crate.

## Shortcut functions
For starters, we can already improve a lot by defining your own shortcut functions. Anyone can already do this for your games. It's also my recommended method since it's the simplest solution and infinitely customizable.

Here's an example of what I'd do.
```rust
fn translate(entity: Entity, start: Vec3, end: Vec3) -> ComponentTween<interpolate::Translation> {
    ComponentTween::new_target(entity, interpolate::Translation { start, end })
}
// or maybe
fn translate_2(
    target: TargetComponent,
    start: Vec3,
    end: Vec3,
) -> ComponentTween<interpolate::Translation> {
    ComponentTween::new_target(target, interpolate::Translation { start, end })
}

fn secs(secs: f32) -> Duration {
    Duration::from_secs_f32(secs)
}

fn tweener(length: Duration) -> SpanTweenerBundle {
    SpanTweenerBundle::new(length)
}

fn setup(mut commands: Commands) {
    let sprite = commands.spawn(SpriteBundle::default()).id();
    commands.spawn(tweener(secs(1.))).with_children(|c| {
        c.span_tweens().tween(
            secs(1.),
            EaseFunction::Linear,
            translate(sprite, Vec3::ZERO, Vec3::ONE),
        );
    });
}
```


## Preset API
[Preset API](https://docs.rs/bevy_tween/0.4.0/bevy_tween/span_tween/struct.SpanTweensBuilder.html#method.add) is also an option.
Use this type alias
```rust
type SpanTweensBuilder<'r, 'a> = bevy_tween::span_tween::SpanTweensBuilder<'r, ChildBuilder<'a>>;
```
if you will only use the child builder but not world child builder. Or make the name even shorter for your liking.

## Keep namespaces short
It's not entirely recommended but if you hate having the `EaseFunction` typed every when then you can use
`use EaseFunction::*;`

## Do whatever you're satisfied with!

### Combinators
Nothing will stop you if you want to go absolutely bongers. This crate is flexible by design and I will make sure you know about it.
Since [SpanTweenPreset](https://docs.rs/bevy_tween/0.4.0/bevy_tween/span_tween/trait.SpanTweenPreset.html#) is implemented for `FnOnce`, any of the combinators down below can be used with the preset APIs.

```rust
fn delay(duration: Duration) -> impl Fn(&mut SpanTweensBuilder) {
    move |b| {
        b.forward(duration);
    }
}

fn repeat<F: Fn(&mut SpanTweensBuilder)>(times: usize, f: F) -> impl Fn(&mut SpanTweensBuilder) {
    move |b| {
        for _ in 0..times {
            f(b);
        }
    }
}

fn tween<I: Bundle, T: Bundle>(
    duration: Duration,
    interpolation: I,
    tween: T,
) -> impl FnOnce(&mut SpanTweensBuilder) {
    move |b| {
        b.tween(duration, interpolation, tween);
    }
}

// this one is clonable to be used in the `repeat` combinator
fn tween_clonable<'a, I: Bundle + Clone, T: Bundle + Clone>(
    duration: Duration,
    interpolation: &'a I,
    tween: &'a T,
) -> impl Fn(&mut SpanTweensBuilder) + 'a {
    move |b| {
        b.tween(duration, interpolation.clone(), tween.clone());
    }
}

macro_rules! seq {
    ($($c:expr,)+) => {
        {
            #[allow(clippy::redundant_closure_call)]
            let c = |b: &mut SpanTweensBuilder| {
                $($c(b);)+
            };
            c
        }
    };
}

fn setup(mut commands: Commands) {
    let sprite = commands.spawn(SpriteBundle::default()).id();
    commands.spawn(tweener(secs(4.))).with_children(|c| {
        seq!(
            tween(
                secs(1.),
                EaseFunction::Linear,
                translate(sprite, Vec3::ZERO, Vec3::ONE)
            ),
            delay(secs(1.)),
            repeat(
                2,
                tween_clonable(
                    secs(1.),
                    &EaseFunction::Linear,
                    &translate(sprite, Vec3::ZERO, Vec3::ONE)
                ),
            ),
        )(&mut c.span_tweens());
    });
}
```

Edit:
Now that I think about this problem for awhile, using combinators might be my best bet. Combinators are powerful and is proven by the `nom` crate. Might give this a try. Would have changes or deprecation on the current builder API.

**Reply by Multirious** [reply]
by @musjj 
> Speaking of verbosity, there might be some lessons that could be learned from https://github.com/nixon-voxell/bevy_motiongfx (and Motion Canvas). I really like how simple and brief the API is:
> 
> ```rust
> chain(&[
>     all(&[
>         commands.play(
>             line_motion
>                 .transform
>                 .translate_add(Vec3::new(0.0, -100.0, 0.0)),
>             1.5,
>         ),
>         commands.play(line_motion.line.extend(100.0), 1.0),
>         commands.play(line_motion.stroke.style_to(10.0), 1.0),
>     ]),
>     all(&[
>         commands.play(
>             line_motion
>                 .transform
>                 .translate_add(Vec3::new(0.0, 100.0, 0.0)),
>             1.5,
>         ),
>         commands.play(line_motion.line.extend(-100.0), 1.0),
>         commands.play(line_motion.stroke.style_to(1.0), 1.0),
>     ]),
> ])
> ```
> 
> So little unnecessary noise which makes constructing complex and long animations a lot easier.
> 
> But after using the plugin for a while, it very much feels that it's optimized for presentation rather than game and interactivity, which is why I ended up moving to `bevy_tween`.

I don't really know much about motiongfx but from looking at the code example you provided, it can be replicated in this crate. Possibly using the combinator style 😙

**Reply by Multirious** [reply]
Update: Common combinators and shortcut functions are now implemented by 606c3d2b717571516d4155c5df9d323cce129e5d

**Reply by musjj** [reply]
This looks very cool, I've been looking at the `improve-ergonomics` branch and it seems very promising :heart_eyes:. I also really like the idea of directly picking up the initial state of the component instead of having to manually specify it. 

Another neat thing with `bevy_motiongfx` is that you can batch set the easings of a group of tweens. Like:

```rust
all(&[
    commands.play(transform_motions[i].translate_add(Vec3::X * 10.0), 1.0),
    commands.play(transform_motions[i].translate_add(Vec3::X * 10.0), 1.0),
])
.with_ease(ease::quart::ease_in_out)
```
What this does is that it sets each tween to the same easing function. Not sure if it'd be possible/makes sense with the current API.

But in any case there are already huge wins in ergonomics here, so I'm pretty happy.

**Reply by Multirious** [reply]
> I also really like the idea of directly picking up the initial state of the component instead of having to manually specify it.

The solution I like to use is custom interpolator like this (from click.rs example):
```rust
    /// Automatically figure out the current position of this entity and then
    /// tween from there.
    pub struct TranslateTo {
        pub start: OnceLock<Vec3>,
        pub end: Vec3,
    }

    impl TranslateTo {
        pub fn end(end: Vec3) -> TranslateTo {
            TranslateTo {
                start: OnceLock::new(),
                end,
            }
        }
    }

    impl Interpolator for TranslateTo {
        type Item = Transform;

        fn interpolate(&self, item: &mut Self::Item, value: f32) {
            let start = self.start.get_or_init(|| item.translation);
            let end = self.end;
            item.translation = start.lerp(end, value);
        }
    }
```
But they feels a bit hacky so I might make something if I thought of something good.

> Another neat thing with bevy_motiongfx is that you can batch set the easings of a group of tweens.

Since we're moving to combinators land now, might as well solves everything with function/closure.
You can do this by caching/currying arguments via a closure. Which works similarly.

``` rust
let translation = |dur, start, end| tween(
    dur,
    EaseFunction::Linear,
    interpolate::translation(start, end)
);

...

builder.add(sequence((
  translation(...),
  translation(...),
)));
```
Or something else.

Make your self familiar with functional programming, they'll be used a lot! The rest of ergonomic improvements from now will likely came from how you utilize functions. They're your tools in this crate!

> But in any case there are already huge wins in ergonomics here, so I'm pretty happy.

Great to hear that! In any case, if there're more ideas or possible improvements, I'll keep updating the crate. These are just my favorite things to update.

**Reply by musjj** [reply]
> But they feels a bit hacky so I might make something if I thought of something good.

If you think it's too hacky, you might want to experiment with how `bevy_motiongfx` is doing it. It doesn't rely on magic, and instead just keeps track of the state in a struct:

```rust
pub struct TransformMotion {
    target_id: Entity,
    transform: Transform,
}
```

Every time you call things like `translate_add`, it mutates the struct and produces an `Action` (similar to this crate's tweens):

```rust
pub fn translate_add(&mut self, translation: Vec3) -> Action<Transform, Vec3, EmptyRes> {
    let new_translation: Vec3 = self.transform.translation + translation;

    let action: Action<Transform, Vec3, EmptyRes> = Action::new(
        self.target_id,
        self.transform.translation, // start
        new_translation, // end
        Self::translation_interp,
    );

    self.transform.translation = new_translation;

    action
}
```

So `TransformMotion` isn't actually involved with the final tweening and just serves as a nice helper to keep track of the state when you're constructing tweens. Not sure which approach is better though.

> Since we're moving to combinators land now, might as well solves everything with function/closure.

I just looked at the examples again, and this looks like what I wanted:

https://github.com/Multirious/bevy_tween/blob/516eff64a3b6c8fd856c260837bc98f68bf4a4e2/examples/demo/event.rs#L50-L62

Doesn't this fulfill the same purpose?

**Reply by Multirious** [reply]
Yes, I'd likely be keeping track of states *somewhere*, that I have to think first where and how yet to maximize usages.

> Doesn't this fulfill the same purpose?

Could you clarify this?

**Reply by musjj** [reply]
Not sure how the function works, but doesn't this sets `EaseFunction::ExponentialIn` to both `interpolate::translation` and `interpolate::angle_z` here?

**Reply by Multirious** [reply]
Oh the function really does nothing special. Both `interpolation` and `tween` argument is of type `Bundle`. They are both insert into the same entity, eg `.spawn((interpolation , tween))`. So yes, those 2 tween would use the same interpolation because they're in the same entity. But they're forced to use the same TimeSpan so there's a bit of limitation.

**Reply by Multirious** [reply]
@musjj What does `transform_motions[i]` gives? Where the array starts from?
Edit: Oh I get it now. It's your custom vector to storing `TransformMotion`. I thought it would be harder to implement. I was thinking like implicit states or something. There's indeed no magic. I'll go for a similar approach.

**Reply by musjj** [reply]
Yeah, it was just a copy-paste from this example: https://github.com/nixon-voxell/bevy_motiongfx/blob/main/examples/easings.rs

**Reply by Multirious** [reply]
Edit: Sorted things up

I came up with many API and not exactly sure which one to select.

# Number 1

```rust
let mut translation = chain(original_translation);
b.add(sequence((
    tween(..., ..., translation.next(translation_to(point1))),
    tween(..., ..., translation.next(translation_to(point2))),
    tween(..., ..., translation.next(translation_to(point3))),
)))
```

Definitions:
```rust
pub fn chain<S>(state: S) -> Chain<S> {
    Chain { state }
}

pub struct Chain<S> {
    state: S,
}

impl<S> Chain<S> {
    pub fn next<E: EntitySpawner, F, O>(&mut self, f: F) -> O
    where
        F: FnOnce(&mut S) -> O,
    {
        f(&mut self.state)
    }
}

pub fn translation_to(
    to: Vec3,
) -> impl Fn(&mut Vec3) -> ComponentTween<Translation> {
    move |v| {
        let start = *v;
        let end = to;
        *v = to;
        translation(start, end)
    }
}
```

# Number 2

```rust
let mut translation = original_translation;
b.add(sequence((
    tween(..., ..., translation_to(point1)(&mut translation)),
    tween(..., ..., translation_to(point2)(&mut translation)),
    tween(..., ..., translation_to(point3)(&mut translation)),
)))
```

# Number 3

```rust
let mut translation = original_translation;
b.add(sequence((
    tween(..., ..., translation_to(&mut translation, point1)),
    tween(..., ..., translation_to(&mut translation, point2)),
    tween(..., ..., translation_to(&mut translation, point3)),
)))
```
Definition
```rust
pub fn translation_to(
    state: &mut Vec3,
    to: Vec3,
) -> ComponentTween<Translation> {
    let start = *state;
    let end = to;
    *state = to;
    translation(start, end)
}
```

# Number 4

```rust
let mut translation_to = translation_to(&mut original_translation);
b.add(sequence((
    tween(..., ..., translation_to(point1)),
    tween(..., ..., translation_to(point2)),
    tween(..., ..., translation_to(point3)),
)))
```
Definition:
```rust
pub fn translation_to(
    state: &mut Vec3,
) -> impl FnMut(Vec3) -> tween::ComponentTween<Translation> + '_ {
    move |to| {
        let start = *state;
        let end = to;
        *state = to;
        translation(start, end)
    }
}
```
This one looks kinda cool but probably out of the question. Mutability issues cause this to be ever usable per one function unless you use interior mutability like `RefCell`.

**Reply by musjj** [reply]
Number 3 looks nice, but I think it would be more natural if `translation_to` is a method of `original_translation` (somehow). So like `original_translation.translation_to(point1)`. Maybe it'd be possible with traits or maybe a wrapper struct.

**Reply by Multirious** [reply]
> Number 3 looks nice, but I think it would be more natural if translation_to is a method of original_translation (somehow). So like original_translation.translation_to(point1). Maybe it'd be possible with traits or maybe a wrapper struct.

What I just worked on! Had t chose Number 1 with a bit of modifications. It came out similar to `bevy_motiongfx`'s `TransformMotion` excepts for combinators and generic. d74ba09

Edit:
I think I'm making this unnecessary complicated. The struct I does on its own is already very simple. Might just go with exactly what `bevy_motiongfx` does. Create new types for every value types. It would be simpler to implement `.do_stuff()` on the new type instead of having a new trait every time or a wrapper struct.

Edit 2:
Oh I think directly adding `.do_stuff()` directly on the interpolator type it self would be very simple. Each interpolator types is *already a wrapper of their values*. Man, I've been on about combinators for so long that a way simpler idea is gone way over my head. ...One caveat that interpolator types don't contain their target. Possible can be resolved with another generic wrapper with some deref abuses for that extra spicy ergonomics.

Edit 3:
Putting state directly on interpolator type will mean that the state is specific to the interpolator and other interpolator that refer to the same value can't use the same state... Back to previous idea 🤭

**Reply by Multirious** [reply]
@musjj  by 13b9a8d, the api is
```rust
use interpolate::transform_tween_state;

let mut transform = transform_tween_state(transform);

b.add(sequence((
    tween(..., ..., transform.translation_to(new_translation)),
    tween(..., ..., transform.scale_to(new_scale)),
    tween(..., ..., transform.rotation_to(new_rotation)),
    tween(..., ..., transform.tween_translation(my_custom_interpolator(...)),
)))
```

**Reply by musjj** [reply]
This looks nice! Btw, would it be possible to mix parallel and sequential tweens with this API? So something like:

```rust
sequence((
    parallel((tween(...), tween(...))),
    sequence((tween(...), tween(...))),
))
```

**Reply by Multirious** [reply]
That's completely fine. All these combinators does is just playing around with the `offset` inside `TweensBuilder`. If the type matched, it will work. I've been making sure that these combinators will work intuitively.

In actual code,
- `sequence` does nothing special because it expects all combinators to shift the offset by their own already. So `sequence` just call all combinators in the tuple in order to emulate sequence.
- `parallel`
  1. saves the current offset
  2. for every combinators inside the tuple elements, use the same saved offset.
  3. The post-`parallel` offset will be the longest offset it has recorded from all combinators in `parallel`.
  4. emulated parallel that also works inside `sequence`!

I might change how they works after this comment but that's how it is for now.

**Reply by musjj** [reply]
Very neat, seems solid to me! Another thought: are tuples the right choice? I'm not sure how it's implemented, but I'm guessing it's done like in bevy (e.g. `impl T for (U, U, U)`) due to the lack of variadics. If so, aren't iterators/lists a better choice so that there's no upper limit?

**Reply by Multirious** [reply]
They're tuples and implemented like Bevy, yep. Iterators will work but they're limited to only 1 type while tuples allows for mix and match any combinators of any types. It's possible to use macro that will allows for infinite elements but rust-analyzer and cargo-fmt is barely working inside them so tuples are the best choice for ergonomics right now. Tuples is implemented upward of 16 elements but it can be nested so you don't have to worry about elements limit.

**Reply by musjj** [reply]
Ok, that makes sense :+1:

**Reply by Multirious** [reply]
Completed by v0.5.0

**Comment by Multirious** [comment]
Maybe have an extension trait on `EntityCommands` to do:
```rust
commands.spawn(/* ... */)
    .with_tweens()
    .tween(/* ... */)
    .tween(/* ... */)
```
for
```rust
commands.spawn(/* ... */)
    .with_children(|c| {
        commands.span_tweens()
            .tween(/* ... */)
            .tween(/* ... */)
    })
```

**Reply by Multirious** [reply]
Update: Now implemented by https://github.com/Multirious/bevy_tween/commit/606c3d2b717571516d4155c5df9d323cce129e5d

**Comment by Multirious** [comment]
New API I want to try
```rust
commands.animate()
    .repeat(Repeat::Infinitely)
    .add(combinators_bla_bla);
```
equals
```rust
commands.spawn(Tweener::new(
    /* automatically resolved time */
).repeat(Repeat::Infinitely)
)
    .tweens()
    .add(combinators_bla_bla);
```

**Reply by Multirious** [reply]
Completed by v0.5.0
